state/accumulator.go

```diff
@@ -333,15 +333,32 @@ func (a *Accumulator) Accumulate(txn *sqlx.Tx, userID, roomID string, timeline i
 	// - there to be no duplicate events
 	// - if there are new events, they are always new.
 	// Both of these assumptions can be false for different reasons
-	dedupedEvents, err := a.filterAndParseTimelineEvents(txn, roomID, timeline.Events, timeline.PrevBatch)
+	dedupedEvents := parseAndDeduplicateTimelineEvents(roomID, timeline)
+	filteredEvents, err := a.filterToNewTimelineEvents(txn, dedupedEvents)
 	if err != nil {
 		err = fmt.Errorf("filterTimelineEvents: %w", err)
 		return
 	}
-	if len(dedupedEvents) == 0 {
+	if len(filteredEvents) == 0 {
 		return 0, nil, err // nothing to do
 	}
 
+	// If this timeline was limited and we don't recognise its first event E3, mark it
+	// as not knowing its previous timeline event.
+	//
+	// We _may_ actually know the previous event. E.g. if the previous poll ends in E1
+	// and we see a limited sync starting with E3, the "gap" may have already been
+	// filled in by other pollers.
+	//
+	//  A:  E1 persisted, E2 omitted and unknown, E3 unknown       (missing previous)
+	//  B:  E1 persisted, E2 omitted and   known, E3 unknown   (not missing previous)
```

Comment:
Terminology is too confusing. Hard to grok what:
 - omitted
 - unknown
 - missing previous

are all supposed to mean. The proxy has either seen the event or not, why do we have 3^2 states?

Reply:
a5c02b2 tries to explain this better.